tools/gbagfx/jasc_pal.c

```diff
@@ -9,58 +9,73 @@
 // Read/write Paint Shop Pro palette files.
 
 // Format of a Paint Shop Pro palette file, line by line:
-// "JASC-PAL\r\n" (signature)
-// "0100\r\n" (version; seems to always be "0100")
-// "<NUMBER_OF_COLORS>\r\n" (number of colors in decimal)
+// "JASC-PAL" (signature)
+// "0100" (version; seems to always be "0100")
+// "<NUMBER_OF_COLORS>" (number of colors in decimal)
 //
 // <NUMBER_OF_COLORS> times:
-// "<RED> <GREEN> <BLUE>\r\n" (color entry)
+// "<RED> <GREEN> <BLUE> <ALPHA (optional)>" (color entry)
+//
+// Line endings can be /r/n or /r or /n
 //
 // Each color component is a decimal number from 0 to 255.
 // Examples:
-// Black - "0 0 0\r\n"
-// Blue  - "0 0 255\r\n"
-// Brown - "150 75 0\r\n"
+// Black - "0 0 0"
+// Blue  - "0 0 255"
+// Brown - "150 75 0"
+// Cyan  - "255 255 255 255"
+
 
-#define MAX_LINE_LENGTH 11
+#define MAX_LINE_LENGTH 15
 
 void ReadJascPaletteLine(FILE *fp, char *line)
 {
     int c;
     int length = 0;
+    int numSpaces = 0; 
 
     for (;;)
     {
         c = fgetc(fp);
 
-        if (c == '\r')
-        {
+        // CRLF or CR
+        if (c == '\r') {
             c = fgetc(fp);
-
-            if (c != '\n')
-                FATAL_ERROR("CR line endings aren't supported.\n");
+            if (c != '\n') {
+                // put it back!
+                ungetc(c, fp);
+            }
 
             line[length] = 0;
-
             return;
         }
 
-        if (c == '\n')
-            FATAL_ERROR("LF line endings aren't supported.\n");
-
-        if (c == EOF)
-            FATAL_ERROR("Unexpected EOF. No CRLF at end of file.\n");
+        // LF only
+        if (c == '\n'){
+            line[length] = 0;
+            return;
+        }
 
         if (c == 0)
-            FATAL_ERROR("NUL character in file.\n");
+            FATAL_ERROR("NULL character in file.\n");
 
-        if (length == MAX_LINE_LENGTH)
-        {
+        if (c == 0)
+            FATAL_ERROR("Unexpected EOF. No CRLF, CR, or LF at end of file.\n");
```

Comment:
The second `FATAL_ERROR` here is unreachable. Additionally, without the error check here for `MAX_LINE_LENGTH` this will now crash on oversized lines.

Reply:
That's a great point, thanks for reviewing this. If the line is over 15 chars but never gets more than 2 spaces, it would've tried to keep reading, i see